### DISCUSSION TITLE
Offical BMW Deezer view for Spotify Integration

### DISCUSSION THREAD

**Discussion by chriss158** [BimmerGestalt/AAIdrive#690] [Ideas]
Hi Walter,

Sadly BMW decided to discontinue the Online Entertainment service (https://f30.bimmerpost.com/forums/showthread.php?t=2002448). Which means that Napster and Deezer cannot be used anymore :(
I liked the Deezer integration by BMW alot. Especially the view of the songs.

![PXL_20230414_103745322](https://user-images.githubusercontent.com/12832850/232223566-0dd58026-1a44-4e5a-8462-65476c15f339.jpg)

Currently the Spotify view is a bit different. It just shows the current song, the previous button and next button.

Would it be possible to clone the view of the deezer integration for spotify? 
Maybe you can also give me a hint where i would have to modify the code to copy the view from deezer.

Thanks in advance.
Best regards
Chris

**Comment by hufman** [comment]
This would be an interesting idea! I know how it would work, but I'm reluctant to do it because Spotify has bugs in its playlist handling, only some of which have workarounds: [Now Playing selection offset](https://github.com/BimmerGestalt/AAIdrive/issues/291), [Artist Playlist selection incorrect](https://github.com/BimmerGestalt/AAIdrive/issues/464), [workaround playlists getting desynced](https://github.com/BimmerGestalt/AAIdrive/issues/398). And of course, some apps don't even provide a playlist, and this playlist navigation screen has less viewable space than the dedicated Now Playing screen.

The ID5+ AudioHmiState represents the middle section of the screen as a list widget, and apps can put whatever information they want in it. The XM and FM radio apps show nearby stations, you see Deezer shows the playlist contents, and the native Bluetooth app shows a Back/Title/Next list [like AAIdrive does](https://github.com/BimmerGestalt/AAIdrive/blob/629fa6751f817c5ca5988818f01be8da4390e163/app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/PlaybackView.kt#L462). It would be straightforward to tweak the [queue view](https://github.com/BimmerGestalt/AAIdrive/blob/main/app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt) to output to this list and route the action handler appropriately.
However, I don't know what UX to implement around this: Should it always try to show a playlist here, if possible? Should the Now Playing button toggle this mode, but how would one get to the fullscreen Now Playing view with beautiful coverart support? Should there be extra complicated logic to ignore buggy Spotify playlists? And so, I'm not eager to try developing this feature.

**Comment by chriss158** [comment]
Thanks for your detailed answer 👍 

It's really a shame that those big streaming services doesn't offer a good API :( Deezer doesn't even allow to show a playlist. Also not for Android Auto. 

> However, I don't know what UX to implement around this: Should it always try to show a playlist here, if possible? Should the Now Playing button toggle this mode, but how would one get to the fullscreen Now Playing view with beautiful coverart support? Should there be extra complicated logic to ignore buggy Spotify playlists? And so, I'm not eager to try developing this feature.

The image i shared is the "Now Playing" view of deezer. So in my mind it should replace the current now playing view, if possible. 

So you are saying if i can somehow manage to show [here ](https://github.com/BimmerGestalt/AAIdrive/blob/629fa6751f817c5ca5988818f01be8da4390e163/app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/PlaybackView.kt#L462) the queue view it should be possible? (Aside from the Spotify bugs)?
If yes, I will try myself at an implementation for this view.

Btw: How do you test your changes? Do you develop something and then go to you car and test your change?

**Comment by hufman** [comment]
It is possible yes, but the code is currently pretty separate and would take a bit of work to route the event handlers and visibility handlers properly.

As for testing: I've become familiar enough with the car API that I write unit tests for the expected API calls (when I do the extra effort to write tests). For smaller changes, I just change the code and test in the car, expecting only a few iterations of running back and forth from my desktop to the car. For debugging, I would use a laptop to deploy and log on the phone while the phone is connected over Bluetooth.
